File: go/nativelib/control.go
package nativelib

import (
	"errors"
	"sync"

	. "github.com/NjinN/RML/go/core"
)

// import "fmt"

func Iif(es *EvalStack, ctx *BindMap) (*Token, error) {
	var args = es.Line[es.LastStartPos() : es.LastEndPos()]

	var result Token
	if args[1].ToBool() {
		if args[2].Tp == BLOCK {
			es.Eval(args[2].Tks(), ctx)
		} else if args[2].Tp == STRING {
			es.EvalStr(args[2].Str(), ctx)
		}
		return &Token{LOGIC, true}, nil
	} else {
		return &Token{NIL, nil}, nil
	}

	result.Tp = ERR
	result.Val = "Type Mismatch"
	return &result, nil

}

func Either(es *EvalStack, ctx *BindMap) (*Token, error) {
	var args = es.Line[es.LastStartPos() : es.LastEndPos()]

	var result Token
	if args[1].ToBool() {
		if args[2].Tp == BLOCK {
			return es.Eval(args[2].Tks(), ctx)
		} else if args[2].Tp == STRING {
			return es.EvalStr(args[2].Str(), ctx)
		}
	} else {
		if args[3].Tp == BLOCK {
			return es.Eval(args[3].Tks(), ctx)
		} else if args[3].Tp == STRING {
			return es.EvalStr(args[3].Str(), ctx)
		}
	}

	result.Tp = ERR
	result.Val = "Type Mismatch"
	return &result, nil
}

func Loop(es *EvalStack, ctx *BindMap) (*Token, error) {
	var args = es.Line[es.LastStartPos() : es.LastEndPos()]

	if args[1].Tp == INTEGER && args[2].Tp == BLOCK {
		var rs *Token
		var err error
		for i := 0; i < args[1].Int(); i++ {
			rs, err = es.Eval(args[2].Tks(), ctx)
			if err != nil {
				if err.Error() == "continue" {
					continue
				}
				if err.Error() == "break" {
					break
				}
				return rs, err
			}
		}

		return rs, nil

	}

	var result = Token{ERR, "Type Mismatch"}
	return &result, nil
}

func Repeat(es *EvalStack, ctx *BindMap) (*Token, error) {
	var args = es.Line[es.LastStartPos() : es.LastEndPos()]
	// args[2].Echo()

	if args[1].Tp == WORD && args[2].Tp == INTEGER && args[3].Tp == BLOCK {
		var c = BindMap{make(map[string]*Token, 8), ctx, TMP_CTX, sync.RWMutex{}}
		var countToken = Token{INTEGER, 1}

		c.PutNow(args[1].Str(), &countToken)
		var rs *Token
		var err error
		for countToken.Int() <= args[2].Int() {
			rs, err = es.Eval(args[3].Tks(), &c)
			countToken.Val = countToken.Int() + 1
			if err != nil {
				if err.Error() == "continue" {
					continue
				}
				if err.Error() == "break" {
					break
				}
				return rs, err
			}
			if rs != nil && rs.Tp == ERR {
				return rs, err
			}
		}
		return &Token{NIL, nil}, nil

	}

	var result = Token{ERR, "Type Mismatch"}
	return &result, nil
}

func Ffor(es *EvalStack, ctx *BindMap) (*Token, error) {
	var args = es.Line[es.LastStartPos() : es.LastEndPos()]

	if args[1].Tp == WORD && args[5].Tp == BLOCK && (args[2].Tp == INTEGER || args[2].Tp == DECIMAL) && (args[3].Tp == INTEGER || args[3].Tp == DECIMAL) && (args[4].Tp == INTEGER || args[4].Tp == DECIMAL) {
		var c = BindMap{make(map[string]*Token, 8), ctx, TMP_CTX, sync.RWMutex{}}
		var countToken = args[2].Dup()
		c.PutNow(args[1].Str(), countToken)
		var rs *Token
		var err error

		if args[2].Tp == INTEGER && args[4].Tp == INTEGER {
			if args[3].Tp == INTEGER {
				for countToken.Int() <= args[3].Int() {
					rs, err = es.Eval(args[5].Tks(), &c)
					countToken.Val = countToken.Int() + args[4].Int()
					if err != nil {
						if err.Error() == "continue" {
							continue
						}
						if err.Error() == "break" {
							break
						}
						return rs, err
					}
					if rs != nil && rs.Tp == ERR {
						return rs, err
					}
				}
			} else {
				for countToken.Int() <= int(args[3].Float()) {
					rs, err = es.Eval(args[5].Tks(), &c)
					countToken.Val = countToken.Int() + args[4].Int()
					if err != nil {
						if err.Error() == "continue" {
							continue
						}
						if err.Error() == "break" {
							break
						}
						return rs, err
					}
					if rs != nil && rs.Tp == ERR {
						return rs, err
					}
				}
			}

		} else if args[2].Tp == INTEGER && args[4].Tp == DECIMAL {
			countToken.Tp = DECIMAL
			countToken.Val = float64(countToken.Int())
			if args[3].Tp == INTEGER {
				for countToken.Float() <= float64(args[3].Int()) {
					rs, err = es.Eval(args[5].Tks(), &c)
					countToken.Val = countToken.Float() + args[4].Float()
					if err != nil {
						if err.Error() == "continue" {
							continue
						}
						if err.Error() == "break" {
							break
						}
						return rs, err
					}
					if rs != nil && rs.Tp == ERR {
						return rs, err
					}
				}
			} else {
				for countToken.Float() <= args[3].Float() {
					rs, err = es.Eval(args[5].Tks(), &c)
					countToken.Val = countToken.Float() + args[4].Float()
					if err != nil {
						if err.Error() == "continue" {
							continue
						}
						if err.Error() == "break" {
							break
						}
						return rs, err
					}
					if rs != nil && rs.Tp == ERR {
						return rs, err
					}
				}
			}
		} else if args[2].Tp == DECIMAL && args[4].Tp == INTEGER {
			if args[3].Tp == INTEGER {
				for countToken.Float() <= float64(args[3].Int()) {
					rs, err = es.Eval(args[5].Tks(), &c)
					countToken.Val = countToken.Float() + float64(args[4].Int())
					if err != nil {
						if err.Error() == "continue" {
							continue
						}
						if err.Error() == "break" {
							break
						}
						return rs, err
					}
					if rs != nil && rs.Tp == ERR {
						return rs, err
					}
				}
			} else {
				for countToken.Float() <= args[3].Float() {
					rs, err = es.Eval(args[5].Tks(), &c)
					countToken.Val = countToken.Float() + float64(args[4].Int())
					if err != nil {
						if err.Error() == "continue" {
							continue
						}
						if err.Error() == "break" {
							break
						}
						return rs, err
					}
					if rs != nil && rs.Tp == ERR {
						return rs, err
					}
				}
			}
		} else if args[2].Tp == DECIMAL && args[4].Tp == DECIMAL {
			if args[3].Tp == INTEGER {
				for countToken.Float() <= float64(args[3].Int()) {
					rs, err = es.Eval(args[5].Tks(), &c)
					countToken.Val = countToken.Float() + args[4].Float()
					if err != nil {
						if err.Error() == "continue" {
							continue
						}
						if err.Error() == "break" {
							break
						}
						return rs, err
					}
					if rs != nil && rs.Tp == ERR {
						return rs, err
					}
				}
			} else {
				for countToken.Float() <= args[3].Float() {
					rs, err = es.Eval(args[5].Tks(), &c)
					countToken.Val = countToken.Float() + args[4].Float()
					if err != nil {
						if err.Error() == "continue" {
							continue
						}
						if err.Error() == "break" {
							break
						}
						return rs, err
					}
					if rs != nil && rs.Tp == ERR {
						return rs, err
					}
				}
			}
		}

		return &Token{NIL, nil}, nil
	}

	var result = Token{ERR, "Type Mismatch"}
	return &result, nil
}

func Wwhile(es *EvalStack, ctx *BindMap) (*Token, error) {
	var args = es.Line[es.LastStartPos() : es.LastEndPos()]

	if args[1].Tp == BLOCK && args[2].Tp == BLOCK {
		var c = BindMap{make(map[string]*Token, 8), ctx, TMP_CTX, sync.RWMutex{}}
		b, e1 := es.Eval(args[1].Tks(), &c)
		if e1 != nil {
			return nil, e1
		}
		var rs *Token
		var err error
		for b.Val.(bool) {
			rs, err = es.Eval(args[2].Tks(), &c)
			if err != nil {
				if err.Error() == "continue" {
					continue
				}
				if err.Error() == "break" {
					break
				}
				return rs, err
			}
			if rs != nil && rs.Tp == ERR {
				return rs, err
			}
			b, err = es.Eval(args[1].Tks(), &c)
			if err != nil {
				return rs, err
			}
			if b.Tp == ERR {
				return rs, err
			}
		}
		return &Token{NIL, nil}, nil
	}
	var result = Token{ERR, "Type Mismatch"}
	return &result, nil
}

func Bbreak(es *EvalStack, ctx *BindMap) (*Token, error) {
	return nil, errors.New("break")
}

func Ccontinue(es *EvalStack, ctx *BindMap) (*Token, error) {
	return nil, errors.New("continue")
}

func Rreturn(es *EvalStack, ctx *BindMap) (*Token, error) {
	var args = es.Line[es.LastStartPos() : es.LastEndPos()]
	return args[1], errors.New("return")
}

func Until(es *EvalStack, ctx *BindMap) (*Token, error) {
	var args = es.Line[es.LastStartPos() : es.LastEndPos()]

	if args[1].Tp ==  BLOCK {
		for {
			rst, err := es.Eval(args[1].Tks(), ctx)

			if err != nil {
				if err.Error() == "break" {
					return &Token{NONE, ""}, nil
				}
				if err.Error() == "continue" {
					continue
				}
				return rst, err
			}

			if rst != nil && rst.ToBool() {
				return rst, err
			}
		}
	}

	var result = Token{ERR, "Type Mismatch"}
	return &result, nil
}


func Ttry(es *EvalStack, ctx *BindMap) (*Token, error) {
	var args = es.Line[es.LastStartPos() : es.LastEndPos()]

	if args[1].Tp == BLOCK && args[2].Tp == BLOCK {
		temp, err := es.Eval(args[1].Tks(), ctx)
		if (temp != nil && temp.Tp == ERR) || err != nil {
			if err != nil {
				temp.Val = err.Error()
			}

			var c = BindMap{make(map[string]*Token, 4), ctx, TMP_CTX, sync.RWMutex{}}
			temp.Tp = STRING
			c.PutNow("e", temp)

			return es.Eval(args[2].Tks(), &c)
		}

		return temp, err
	}

	return &Token{ERR, "Type Mismatch"}, nil
}

func Cause(es *EvalStack, ctx *BindMap) (*Token, error) {
	var args = es.Line[es.LastStartPos() : es.LastEndPos()]

	if args[1].Tp == STRING {
		return &Token{ERR, args[1].Str()}, nil
	}

	return &Token{ERR, "Type Mismatch"}, nil
}

File: other-lang-version/nim-stack/native/control.nim
proc iff*(stack: var EvalStack, ctx: var BindMap[Token]):Token=
    var args = addr(stack.line[stack.startPos.last])
    if args[2].tp == TypeEnum.list:
        case args[1].tp
        of TypeEnum.logic:
            if args[1].val.logic:
                result = stack.eval(args[2].val.list, ctx)
                return result
            else:
                return initToken(TypeEnum.null)
        of TypeEnum.integer:
            if args[1].val.integer != 0:
                result = stack.eval(args[2].val.list, ctx)
                return result
            else:
                return initToken(TypeEnum.null)
        of TypeEnum.decimal:
            if args[1].val.decimal != 0.0:
                result = stack.eval(args[2].val.list, ctx)
                return result
            else:
                return initToken(TypeEnum.null)
        of TypeEnum.string:
            if args[1].val.string != "":
                result = stack.eval(args[2].val.list, ctx)
                return result
            else:
                return initToken(TypeEnum.null)
        of TypeEnum.none:
            return initToken(TypeEnum.null)
        else:
            result = stack.eval(args[2].val.list, ctx)
            return result
    else:
        result = initToken(TypeEnum.err)
        result.val.string = "Type Mismatch"
        return result


proc either*(stack: var EvalStack, ctx: var BindMap[Token]):Token=
    var args = addr(stack.line[stack.startPos.last])
    if args[2].tp == TypeEnum.list and args[3].tp == TypeEnum.list:
        case args[1].tp
        of TypeEnum.logic:
            if args[1].val.logic:
                result = stack.eval(args[2].val.list, ctx)
            else:
                result = stack.eval(args[3].val.list, ctx)
        of TypeEnum.integer:
            if args[1].val.integer != 0:
                result = stack.eval(args[2].val.list, ctx)
            else:
                result = stack.eval(args[3].val.list, ctx)
        of TypeEnum.decimal:
            if args[1].val.decimal != 0.0:
                result = stack.eval(args[2].val.list, ctx)
            else:
                result = stack.eval(args[3].val.list, ctx)
        of TypeEnum.string:
            if args[1].val.string != "":
                result = stack.eval(args[2].val.list, ctx)
            else:
                result = stack.eval(args[3].val.list, ctx)
        of TypeEnum.none:
            result = stack.eval(args[3].val.list, ctx)
        else:
            result = stack.eval(args[2].val.list, ctx)
    else:
        result = initToken(TypeEnum.err)
        result.val.string = "Type Mismatch"
    return result


proc loop*(stack: var EvalStack, ctx: var BindMap[Token]):Token=
    var args = addr(stack.line[stack.startPos.last])
    if args[1].tp == TypeEnum.integer and args[2].tp == TypeEnum.list:
        # print args[1]
        # var s = readLine(stdin)
        for i in 1..args[1].val.integer:
            try:
                result = stack.eval(args[2].val.list, ctx)
            except:
                if getCurrentExceptionMsg() == "continue":
                    continue
                elif getCurrentExceptionMsg() == "break":
                    break
                else:
                    raise getCurrentException()

    else:
        result = initToken(TypeEnum.err)
        result.val.string = "Type Mismatch"
    return result

proc repeat*(stack: var EvalStack, ctx: var BindMap[Token]):Token=
    var args = addr(stack.line[stack.startPos.last])
    if args[1].tp == TypeEnum.word and args[2].tp == TypeEnum.integer and args[3].tp == TypeEnum.list:
        var c = initBindMap[Token](4)
        c.father = addr(ctx)
        var countToken = initToken(TypeEnum.integer)
        
        countToken.val.integer = 1
        c[args[1].val.string] = countToken 
        while c[args[1].val.string].val.integer <= args[2].val.integer:
            try:
                result = stack.eval(args[3].val.list, c)
            except:
                if getCurrentExceptionMsg() == "continue":
                    continue
                elif getCurrentExceptionMsg() == "break":
                    break
                else:
                    raise getCurrentException()
            finally:
                # echo(repr(unit.mainCtx[args[1].val.string]))
                var temp = c[args[1].val.string]
                temp.val.integer += 1
                c[args[1].val.string] = temp
        freeBindMap(c)
    else:
        result = initToken(TypeEnum.err)
        result.val.string = "Type Mismatch"
    return result


proc ffor*(stack: var EvalStack, ctx: var BindMap[Token]):Token=
    var args = addr(stack.line[stack.startPos.last])
    if args[1].tp == TypeEnum.word and args[5].tp == TypeEnum.list and (args[2].tp == TypeEnum.integer or args[2].tp == TypeEnum.decimal) and (args[3].tp == TypeEnum.integer or args[3].tp == TypeEnum.decimal) and (args[4].tp == TypeEnum.integer or args[4].tp == TypeEnum.decimal) :
        var c = initBindMap[Token](4)
        c.father = addr(ctx)
        if(args[2].tp == TypeEnum.integer and args[3].tp == TypeEnum.integer and args[4].tp == TypeEnum.integer):
            var count = initToken(TypeEnum.integer)
            
            count.val.integer = args[2].val.integer
            c[args[1].val.string] = count
            while c[args[1].val.string].val.integer <= args[3].val.integer:
                try:
                    result = stack.eval(args[5].val.list, c)
                except:
                    if getCurrentExceptionMsg() == "continue":
                        continue
                    elif getCurrentExceptionMsg() == "break":
                        break
                    else:
                        raise getCurrentException()
                finally:
                    var temp = c[args[1].val.string]
                    temp.val.integer += args[4].val.integer
                    c[args[1].val.string] = temp
            
        else:
            var count = initToken(TypeEnum.decimal)
            
            if args[2].tp == TypeEnum.integer:
                count.val.decimal = float64(args[2].val.integer)
            else:
                count.val.decimal = args[2].val.decimal
            c[args[1].val.string] = count
            
            var temp: Token
            if args[3].tp == TypeEnum.integer:
                while c[args[1].val.string].val.decimal <= args[3].val.integer.float64:
                    try:
                        result = stack.eval(args[5].val.list, c)
                    except:
                        if getCurrentExceptionMsg() == "continue":
                            continue
                        elif getCurrentExceptionMsg() == "break":
                            break
                        else:
                            raise getCurrentException()
                    finally:
                        if args[4].tp == TypeEnum.integer:
                            temp = c[args[1].val.string]
                            temp.val.decimal += args[4].val.integer.float64
                            c[args[1].val.string] = temp
                        else:
                            temp = c[args[1].val.string]
                            temp.val.decimal += args[4].val.decimal
                            c[args[1].val.string] = temp
            else:
                while c[args[1].val.string].val.decimal <= args[3].val.decimal:
                    try:
                        result = stack.eval(args[5].val.list, ctx)
                    except:
                        if getCurrentExceptionMsg() == "continue":
                            continue
                        elif getCurrentExceptionMsg() == "break":
                            break
                        else:
                            raise getCurrentException()
                    finally:
                        if args[4].tp == TypeEnum.integer:
                            temp = c[args[1].val.string]
                            temp.val.decimal += args[4].val.integer.float64
                            c[args[1].val.string] = temp
                        else:
                            temp = c[args[1].val.string]
                            temp.val.decimal += args[4].val.decimal
                            c[args[1].val.string] = temp
            
        freeBindMap(c)
    else:
        result = initToken(TypeEnum.err)
        result.val.string = "Type Mismatch"
    return result


proc wwhile*(stack: var EvalStack, ctx: var BindMap[Token]):Token=
    var args = addr(stack.line[stack.startPos.last])
    if args[1].tp == TypeEnum.list and args[2].tp == TypeEnum.list:
        var condCtx = initBindMap[Token](4)
        condCtx.father = addr(ctx)
        var bodyCtx = initBindMap[Token](4)
        bodyCtx.father = addr(ctx)
        var b = stack.eval(args[1].val.list, condCtx)
        
        while b.tp == TypeEnum.logic and b.val.logic:
            try:
                result = stack.eval(args[2].val.list, bodyCtx)
            except:
                if getCurrentExceptionMsg() == "continue":
                    continue
                elif getCurrentExceptionMsg() == "break":
                    break
                else:
                    raise getCurrentException()
            finally:
                b = stack.eval(args[1].val.list, condCtx)

                if not (b.tp == TypeEnum.logic):
                    result = initToken(TypeEnum.err)
                    result.val.string = "Bad Logic Expression "
                    return result
        freeBindMap(condCtx)
        freeBindMap(bodyCtx)
    else:
        result = initToken(TypeEnum.err)
        result.val.string = "Type Mismatch"
    return result

proc bbreak*(stack: var EvalStack, ctx: var BindMap[Token]):Token=
    raise newException(CatchableError, "break")

proc ccontinue*(stack: var EvalStack, ctx: var BindMap[Token]):Token=
    raise newException(CatchableError, "continue")